provider/pkg/provider/provider.go

```diff
@@ -1077,6 +1102,30 @@ func (p *cfnProvider) Update(ctx context.Context, req *pulumirpc.UpdateRequest)
 	var outputs resource.PropertyMap
 	id := req.GetId()
 	resourceToken := string(urn.Type())
+
+	if req.GetPreview() {
+		if _, ok := p.customResources[resourceToken]; ok {
+			outputs = PreviewOutputs(newInputs, nil, nil, nil)
+		} else {
+			spec, hasSpec := p.resourceMap.Resources[resourceToken]
+			if !hasSpec {
+				return nil, errors.Errorf("Resource type %s not found", resourceToken)
+			}
+			resourceTypeName := urn.Type().Name()
+			outputs = PreviewOutputs(newInputs, p.resourceMap.Types, spec.Outputs, spec.ReadOnly)
+			outputs = populateStableOutputs(newInputs, oldState, spec.ReadOnly, resourceTypeName)
```

Comment:
Why is populateStableOutputs not a part of PreviewOutputs implementation?

Reply:
Mainly because we want to use `PreviewOutputs` on both `Create` & `Update`, but we only use `populateStableOutputs` on `Update`. If you think it would be better to combine it all in `PreviewOutputs` and just decide what to do  based on whether `oldState` is provided I could do that as well, I don't have a strong preference either way.

Reply:
Neither do I, feel free to ignore, it is a matter of style, this organization also makes sense. Just looked a little weird to me (also, different capitalization, is PreviewOutputs exported but populateStableOutputs is not)?